romanisim/tests/test_l1.py

```diff
@@ -84,6 +84,39 @@ def test_apportion_counts_to_resultants():
     log.info('DMS229: successfully generated ramp from counts.')
 
 
+def test_linearized_counts_to_resultants():
+    """Test that we can apportion linearized counts to resultants.
+    """
+    counts = np.random.poisson(100, size=(100, 100))
+    coeffs = np.asfarray([1.0, 0.7, 3.0e-6, 5.0e-12])
+    lin_coeffs = np.tile(coeffs[:,np.newaxis,np.newaxis], (1, 100, 100))
+    lin_coeffs[:, 0:50, :] *= 2.0
+    gain = 4.0
+    counts[0,0] = counts[0,99] = counts[99,0] = counts[99,99] = 11.0
+
+    linearity = nonlinearity.NL(lin_coeffs, gain)
+
+    for tij in tijlist:
+        resultants, dq = l1.apportion_counts_to_resultants(counts.copy(), tij.copy())
+        assert np.all(np.diff(resultants, axis=0) >= 0)
+        assert np.all(resultants >= 0)
+        assert np.all(resultants <= counts[None, :, :])
+
+        res2, dq2 = l1.apportion_counts_to_resultants(counts.copy(), tij.copy(), linearity=linearity)
+
+        assert np.all(res2 >= 0)
+
+        assert np.all(res2 != resultants)
+
+        assert np.all(res2[-1:1, 0, 0] != res2[-1:1, 0, 99])
+        assert np.all(res2[:, 0, 0] != res2[:, 99, 0])
+        assert np.all(res2[:, 0, 0] != res2[:, 99, 99])
+        assert np.all(res2[:, 0, 99] != res2[:, 99, 0])
+        assert np.all(res2[:, 0, 99] != res2[:, 99, 99])
+        assert np.all(res2[:-1, 99, 0] != res2[:-1, 99, 99])
```

Comment:
It's not true that any of these need to be true for any possible set of random numbers or tij?  e.g. for the second one, res2[:, 0, 0] compares the resultants for one corner with the ones for another.  You gave them the same total number of counts.  You gave them pretty different linearity coefficients, but if the first resultant in the one corner happens to get more / less photons early on, that could lead to a match?

It looks to me like I'm expecting the first coefficient in the file to be around 0 and the second to be around 1, and the higher coefficients to be small.  Your example coefficients don't do that.  Is that intentional?

Some tests that you could do that would make more sense to me:
- generate res, res2 with the same random number generator.
- give res2 some simple linearity coefficients, e.g., [0, 0.5, 0, ...].  This effectively changes the gain by a factor of 2.
- check that np.all(res2 <= res) and np.any(res2 < res), which should be ~guaranteed (at least one resultant got fewer counts (since a count now corresponds to two electrons), and all of them are less than or equal.

I agree with you that we also want a simple test that the coefficients and their inverse invert.  e.g., coeffs = [0, 2, 0, 0].  coeffs2 = [0, 0.5, 0, 0].  nl1.apply(nl2.apply(values)) == values.